flaky: `chainer_tests/serializers_tests/test_hdf5.py::TestHDF5Deserializer::test_deserialize_different_dtype_cpu`

https://jenkins.preferred.jp/job/chainer/job/chainer_pr/2114/TEST=chainer-py2,label=mn1-p100/console

`18:57:22 FAIL tests/chainer_tests/serializers_tests/test_hdf5.py::TestHDF5Deserializer::test_deserialize_different_dtype_cpu`

```
18:57:21 E                   AssertionError: 
18:57:21 E                   Arrays are not equal
18:57:21 E                   
18:57:21 E                   (mismatch 16.6666666667%)
18:57:21 E                    x: array([[ 0.272705, -0.590332,  0.962891],
18:57:21 E                          [-0.482178, -0.544922, -0.029999]], dtype=float16)
18:57:21 E                    y: array([[ 0.272705, -0.590332,  0.962891],
18:57:21 E                          [-0.481934, -0.544922, -0.029999]], dtype=float16)
```

--- a/tests/chainer_tests/serializers_tests/test_hdf5.py
+++ b/tests/chainer_tests/serializers_tests/test_hdf5.py
@@ -5,6 +5,7 @@ import unittest
 
 import mock
 import numpy
+import six
 
 import chainer
 from chainer import backend
@@ -181,8 +182,23 @@ class TestHDF5Deserializer(unittest.TestCase):
     def test_deserialize_different_dtype_cpu(self):
         y = numpy.empty((2, 3), dtype=numpy.float16)
         ret = self.deserializer('y', y)
-        numpy.testing.assert_array_equal(y, self.data.astype(numpy.float16))
+
         self.assertIs(ret, y)
+
+        # Compare the value with the original array. Note that it's not always
+        # bit-identical to the result of numpy.ndarray.astype.
+        numpy.testing.assert_allclose(
+            y, self.data.astype(numpy.float16),
+            rtol=1e-3, atol=1e-3)
+
+        # It should be bit-identical to the result directly retrieved from
+        # h5py.
+        arr_hdf5 = numpy.empty((2, 3), dtype=numpy.float16)
+        buf = six.BytesIO()
+        with h5py.File(buf, 'w') as f:
+            f.create_dataset('a', data=self.data)
+            f['a'].read_direct(arr_hdf5)
+        numpy.testing.assert_array_equal(y, arr_hdf5)
 
     @attr.gpu
     def test_deserialize_different_dtype_gpu(self):

--- a/tests/chainer_tests/serializers_tests/test_hdf5.py
+++ b/tests/chainer_tests/serializers_tests/test_hdf5.py
@@ -5,7 +5,6 @@ import unittest
 
 import mock
 import numpy
-import six
 
 import chainer
 from chainer import backend
@@ -194,10 +193,10 @@ class TestHDF5Deserializer(unittest.TestCase):
         # It should be bit-identical to the result directly retrieved from
         # h5py.
         arr_hdf5 = numpy.empty((2, 3), dtype=numpy.float16)
-        buf = six.BytesIO()
-        with h5py.File(buf, 'w') as f:
-            f.create_dataset('a', data=self.data)
-            f['a'].read_direct(arr_hdf5)
+        with tempfile.TemporaryFile() as buf:
+            with h5py.File(buf, 'w') as f:
+                f.create_dataset('a', data=self.data)
+                f['a'].read_direct(arr_hdf5)
         numpy.testing.assert_array_equal(y, arr_hdf5)
 
     @attr.gpu

